File: src/grid/grid.ts
import { makeSquare } from "../utils/grid.utils.ts";
import { FindPathConfig, Point } from "../types/main.ts";
import { findPath } from "./finder.ts";

const NON_WALKABLE_HEIGHT = 0;

export class Grid {
  public readonly width: number;
  public readonly height: number;
  private readonly heightMatrix: Float32Array;

  constructor(width: number, height: number, costMatrix: Float32Array) {
    this.width = width;
    this.height = height;
    this.heightMatrix = costMatrix;

    if (width !== height) throw new Error("grid matrix must be square");

    if (costMatrix.length !== width * height)
      throw new Error("cost matrix must have the same dimensions as the grid");
  }

  public static from(matrix: number[][]): Grid {
    if (matrix[0] === undefined || matrix[0][0] === undefined)
      throw new Error("grid matrix cannot be empty");

    const mat = makeSquare(matrix);
    const height = mat.length;
    const width = mat[0].length;

    if (height !== width) throw new Error("grid matrix must be square");

    const costMatrix = new Float32Array(height * width);
    mat.forEach((row, y) => {
      row.forEach((cost, x) => {
        costMatrix[y * width + x] = cost;
      });
    });

    return new Grid(width, height, costMatrix);
  }

  public getHeightAt(point: Point): number | null {
    if (!this.inBounds(point)) return null;
    const cost = this.heightMatrix[this.index(point)];
    return cost === NON_WALKABLE_HEIGHT ? null : cost;
  }

  public distance(a: Point, b: Point): number {
    // return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
    return Math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2);
  }

  public index(point: Point): number {
    return point.y * this.height + point.x;
  }

  private clone(): Grid {
    return new Grid(
      this.width,
      this.height,
      new Float32Array(this.heightMatrix),
    );
  }

  public inBounds(point: Point): boolean {
    return (
      point.x >= 0 &&
      point.x < this.width &&
      point.y >= 0 &&
      point.y < this.height
    );
  }

  public isWalkable(point: Point): boolean {
    const heightAt = this.getHeightAt(point);
    return (
      this.inBounds(point) &&
      heightAt !== null &&
      heightAt !== NON_WALKABLE_HEIGHT
    );
  }

  public walkMatrix(
    callback: (x: number, y: number, cost: number | null) => void,
  ) {
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        callback(x, y, this.heightMatrix[y * this.height + x]);
      }
    }
  }

  public mapMatrix<T>(
    callback: (x: number, y: number, cost: number | null) => T,
  ): T[][] {
    const matrix: T[][] = [];

    for (let y = 0; y < this.height; y++) {
      const row: T[] = [];
      for (let x = 0; x < this.width; x++) {
        const value: T = callback(x, y, this.heightMatrix[y * this.height + x]);
        row.push(value);
      }
      matrix.push(row);
    }

    return matrix;
  }

  public findPath(
    startPoint: Point,
    endPoint: Point,
    config: FindPathConfig = {
      maxJumpCost: 5,
      travelCosts: undefined,
    },
  ): Point[] {
    return findPath(startPoint, endPoint, this, config);
  }
}

File: src/grid.rs
use crate::point::Point;
use crate::utils::make_square;

#[derive(Debug)]
pub struct Grid {
    pub width: usize,
    pub height: usize,
    pub height_matrix: Vec<f32>,
}

impl Grid {
    pub fn new(width: usize, height: usize, cost_matrix: Vec<f32>) -> Self {
        Grid {
            width,
            height,
            height_matrix: cost_matrix,
        }
    }

    pub fn from(matrix: Vec<Vec<usize>>) -> Result<Self, String> {
        if matrix.get(0).is_none() || matrix.get(0).unwrap().get(0).is_none() {
            return Err("Grid matrix cannot be empty!".to_string());
        }

        let mat = make_square(matrix);
        let height = mat.len();
        let width = mat.get(0).unwrap().len();

        if height != width {
            return Err("Grid matrix must be square!".to_string());
        }

        let mut cost_matrix = vec![0.0; width * height];

        for (y, row) in mat.iter().enumerate() {
            for (x, &cost) in row.iter().enumerate() {
                cost_matrix[y * width + x] = cost as f32
            }
        }

        Ok(Grid::new(width, height, cost_matrix))
    }

    pub fn get_height_at(&self, point: &Point) -> Option<f32> {
        if !self.in_bounds(&point) {
            return None;
        }
        Some(self.height_matrix[self.index(point)])
    }
    pub fn distance(&self, point_a: &Point, point_b: &Point) -> usize {
        let x_expo = i32::pow((point_a.x - point_b.x) as i32, 2);
        let y_expo = i32::pow((point_a.y - point_b.y) as i32, 2);

        i32::isqrt(x_expo + y_expo) as usize
    }

    pub fn index(&self, point: &Point) -> usize {
        (point.y * self.height as isize + point.x) as usize
    }

    pub fn in_bounds(&self, point: &Point) -> bool {
        point.x >= 0
            && point.x < self.width as isize
            && point.y >= 0
            && point.y < self.height as isize
    }

    pub fn is_walkable(&self, point: &Point) -> bool {
        let height_at = self.get_height_at(&point);
        self.in_bounds(point)
	        && height_at.is_some()
	        //NON WALKABLE HEIGHT
	        && height_at.unwrap() != 0f32
    }

    pub fn walk_matrix<F>(&self, mut callback: F)
    where
        F: FnMut(usize, usize, Option<f32>),
    {
        for y in 0..self.height {
            for x in 0..self.width {
                let index = y * self.height + x;
                let cost = self.height_matrix[index];
                callback(x, y, Some(cost));
            }
        }
    }
}
